libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/util/Geocoder.kt

```diff
@@ -0,0 +1,56 @@
+package com.mapbox.navigation.dropin.util
+
+import com.mapbox.api.geocoding.v5.GeocodingCriteria
+import com.mapbox.api.geocoding.v5.GeocodingCriteria.GeocodingTypeCriteria
+import com.mapbox.api.geocoding.v5.MapboxGeocoding
+import com.mapbox.api.geocoding.v5.models.CarmenFeature
+import com.mapbox.api.geocoding.v5.models.GeocodingResponse
+import com.mapbox.geojson.Point
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withContext
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+/**
+ * Wrapper class for [MapboxGeocoding].
+ */
+internal class Geocoder(
+    private val accessToken: String,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+    suspend fun findAddresses(point: Point): List<CarmenFeature> =
+        reverseGeocode(point, GeocodingCriteria.TYPE_ADDRESS)?.features() ?: emptyList()
+
+    suspend fun reverseGeocode(
+        point: Point,
+        @GeocodingTypeCriteria criteria: String
+    ): GeocodingResponse? = withContext(dispatcher) {
+        suspendCancellableCoroutine { continuation ->
+            val reverseGeocode = MapboxGeocoding.builder()
+                .accessToken(accessToken)
+                .query(point)
+                .geocodingTypes(criteria)
+                .build()
+
+            reverseGeocode.enqueueCall(object : Callback<GeocodingResponse> {
+                override fun onResponse(
+                    call: Call<GeocodingResponse>,
+                    response: Response<GeocodingResponse>
+                ) {
+                    continuation.resume(response.body())
+                }
+
+                override fun onFailure(call: Call<GeocodingResponse>, throwable: Throwable) {
+                    continuation.resumeWithException(throwable)
```

Comment:
Instead of throwing an exception return an Expected.

Reply:
I agree to this. Let's return an Expected as we do so across our entire SDK to maintain consistency

Reply:
Updated. But for the record I disagree with both of you. Exceptions are correct error handling mechanism here. They allow "bubbling up" exceptions to a high layers while Result/Expected don't.

Reply:
Returning an Expected makes the method deterministic and it's the convention we're following in the SDK.  In addition an Expected<Error> will pass the error up to higher layers in case additional action is needed based on the error.

Reply:
Yes agreed with @cafesilencio. I would be in favor of adopting the same conventions we have been in our SDK. We can still define our own behaviors based on the state of Expected and the error it contains.

Reply:
@tomaszrybakiewicz Are you planning to fix it. We should also add some way to be able to cancel this request in case it's ongoing and `GeocodingComponent` gets detached.

Reply:
This method is private and I am not planning on changing it anymore. If you think you can improve it, go ahead and change it.